src/doc/performance-tuning.md

```diff
@@ -1,91 +1,185 @@
 # Maximizing deployment performance
 
-For maximum performance, several hardware, operating system and Snabb  parameters need
-to be tuned. Note that this document is only on tuning for deployment
-performance, not on how to write performant Snabb code.
+For maximum performance, several hardware, operating system and Snabb
+parameters need to be tuned.  We start with some generic
+system-oriented tips and move towards application configuration.
 
-## Snabb
-### ring-buffer/num_descriptors
-Defined in src/apps/intel10g.lua and adjustable before (!) the NIC gets initialized:
+## BIOS settings
 
-```
-require('apps.intel.intel10g').num_descriptors = ring_buffer_size
-...
-config.app(c, "nic", require(device_info.driver).driver, {...})
+### Prioritize performance over efficiency
+
+Go into your BIOS and verify that you or your hardware vendor have not
+enabled aggressive power-saving modes that could downclock your
+processors.  A CPU in a power-saving mode typically takes some time to
+return to peak performance, and this latency can cause packet loss.
+Servers are often shipped configured to prioritize energy efficiency
+over performance, which is not the best choice if avoiding packet loss
+is a goal.
+
+Note that each vendor offers different BIOS settings to enable or
+disable energy efficiency settings or profiles, typically named "Max
+performance", "Energy Efficiency" and "Custom".  Select "Max
+performance" for latency sensitive Snabb use.
+
+### Disable hyperthreads
+
+Hyperthreads are a way of maximizing resource utilization on a CPU core,
+driven by the observation that a CPU is often waiting on memory or some
+external event, and might as well be doing something else while it's
+waiting.  In such a situation, it can be advantageous to run a second
+thread on that CPU.  However for Snabb that's exactly what we don't
+want.  We do not want another thread competing for compute and cache
+resources on our CPU and increasing our latency.  For best results and
+lowest latency, disable hyperthreading via the BIOS settings.
+
```

Comment:
Having thought a bit more, I'd really like this to explicitly say that we have observed hyperthreading causing packet loss.